src/future/index.js

```diff
@@ -0,0 +1,108 @@
+import React, { useState, useMemo } from 'react';
+
+import useField from '@/src/future/hooks/useField';
+
+import { FieldContext, RenderContext } from '@/src/future/context';
+
+import Component from '@/src/future/components/Component';
+import Field from '@/src/future/components/Field';
+
+import TextInput from '@/src/future/inputs/TextInput';
+
+function ControlledFormContainer({ value, onChange, children }) {
+  function onSetFieldValue(fieldKey, fieldValue) {
+    const newValue = {
+      ...value,
+      [fieldKey]: fieldValue,
+    };
+    onChange(newValue);
+  }
+  return (
+    <FieldContext.Provider value={{ formValue: value, onSetFieldValue }}>
+      {children}
+    </FieldContext.Provider>
+  );
+}
+
+function UncontrolledFormContainer({ defaultValue, onChange, children }) {
+  const [formValue, setFormValue] = useState(defaultValue);
+
+  function onChangeControlled(newValue) {
+    setFormValue(newValue);
+    onChange(newValue);
+  }
+
+  return (
+    <ControlledFormContainer onChange={onChangeControlled} value={formValue}>
+      {children}
+    </ControlledFormContainer>
+  );
+}
+
+export function FormContainer({
+  value,
+  defaultValue,
+  renderTag,
+  renderComponent,
+  ...props
+}) {
+  const [savedDefaultValue] = useState(defaultValue);
+  const [isControlled] = useState(savedDefaultValue === undefined);
```

Comment:
Do we need to put this in state? I could see caching the initial value of `props.defaultValue`.

Reply:
That's effectively what I'm doing here. When you use `useState` without changing the value, it's effectively like using a constructor in a class-based component. I'm saving the value for later rather than mutating it every time. The value passed to `useState` is the initial value and will never change here.